File: src/NodeParser/UndefinedLiteralNodeParser.ts
import ts from "typescript";
import type { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { NullType } from "../Type/NullType.js";

export class UndefinedLiteralNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.UndefinedKeyword;
    }
    public createType(node: ts.KeywordTypeNode, context: Context): BaseType {
        return new NullType();
    }
}

File: internal/parser/undefined_literal_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// UndefinedLiteralNodeParser maps the `undefined` keyword to NullType.
// Like the original (src/NodeParser/UndefinedLiteralNodeParser.ts), it is
// not registered in the default parser chain; UndefinedTypeNodeParser wins.
type UndefinedLiteralNodeParser struct{}

func NewUndefinedLiteralNodeParser() *UndefinedLiteralNodeParser {
	return &UndefinedLiteralNodeParser{}
}

func (p *UndefinedLiteralNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindUndefinedKeyword
}

func (p *UndefinedLiteralNodeParser) CreateType(_ *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return &types.NullType{}
}
